### DISCUSSION TITLE
Granular permissions

### DISCUSSION THREAD

**Discussion by Wulfheart** [filamentphp/filament#5470] [Ideas]
Currently it is only possible to make whole resources editable by creating policies. However, I would like to make different aspects editable (example below, I might need to add some `unset` here). An admin should be allowed to edit something but a normal user just the description, etc.
So I figure it might look something like this:
```php
Forms\Components\TextInput::make('name')
    ->allowEdit(true)
    // or the inverse
    ->prohibitEdit(fn() => false)
```

What do you think about it? Is this something I could also PR into v2? Are there already plans for this for v3?

### Example
![image](https://user-images.githubusercontent.com/25671390/212478802-8e293a3c-f529-4e3c-8e28-a19f791fd5fe.png)
```php
return $form
    ->schema([
        Forms\Components\TextInput::make('name')
            ->label(__('core/game.attributes.name'))
            ->required()
            ->disabledOn('edit')
            ->unique(ignoreRecord: true),
        Forms\Components\Select::make('variant_id')
            ->label(__('core/variant.name_singular'))
            ->relationship('variant', 'name')
            ->required()
            ->disabledOn('edit')
            ->dehydrated(fn (Page $livewire) => $livewire instanceof CreateRecord),
        Forms\Components\MarkdownEditor::make('description')
            ->label(__('core/game.attributes.description'))
            ->disableToolbarButtons([
                'attachFiles',
                'codeBlock',
            ])
            ->columnSpanFull(),
    ]);
```

**Comment by danharrin** [comment]
You can already use `disabled(fn () => check perms)` etc

**Reply by Wulfheart** [reply]
Yes. But this seems tedious to me. Would you be willing to merge such a trait? Is there any other way to supply it to the current project without changing filament?
Is there something like a query builder where I can define the zwei statements once?

**Reply by danharrin** [reply]
sorry, I don't understand the difference between `prohibitEdit()` and `disabled()`

**Reply by Wulfheart** [reply]
Is it like hidden where there is special protection in place if it disabled is true?

**Reply by danharrin** [reply]
In v3, yes. In v2, you need to pair it with `dehydrated()`

**Reply by Wulfheart** [reply]
Ah. Ok. Thanks. That clarifies it for me

**Comment by ArtMin96** [comment]
Is this not what you want?
```php
TextInput::make('name')->hidden(! auth()->user()->isAdmin())
```
For action
```php
Action::make('edit')
    ->url(fn (Post $record): string => route('posts.edit', $record))
    ->visible(fn (Post $record): bool => auth()->user()->can('update', $record))
```
